src/Framework/Constraint/Count.php

```diff
@@ -56,64 +62,86 @@ protected function getCountOf($other): ?int
         }
 
         if ($other instanceof Traversable) {
-            while ($other instanceof IteratorAggregate) {
-                $other = $other->getIterator();
-            }
-
-            $iterator = $other;
+            return $this->getCountOfTraversable($other);
+        }
+    }
 
-            if ($iterator instanceof Generator) {
-                return $this->getCountOfGenerator($iterator);
-            }
+    /**
+     * Returns the description of the failure.
+     *
+     * The beginning of failure messages is "Failed asserting that" in most
+     * cases. This method should return the second part of that sentence.
+     *
+     * @param mixed $other evaluated value or object
+     */
+    protected function failureDescription($other): string
+    {
+        return \sprintf(
+            'actual size %d matches expected size %d',
+            $this->getCountOf($other),
+            $this->expectedCount
+        );
+    }
 
-            if (!$iterator instanceof Iterator) {
-                return \iterator_count($iterator);
-            }
+    private function getCountOfTraversable(Traversable $traversable): int
+    {
+        if ($this->iteratorCounts === null) {
+            $this->iteratorCounts = new \SplObjectStorage();
+        }
```

Comment:
I do not understand how without this extra initialization the field can end up being null. Yet that is what apparently happens: https://travis-ci.org/sebastianbergmann/phpunit/jobs/434909673

How is this possible? It is set in the constructor. SameSize is calling `parent::__construct`. Nothing besides the constructor writes to the field and the field is private. Huh???